### PR TITLE
Fix todo: avoid relying on `logits_all == true` in perplexity_v2

### PR DESCRIPTION
**Pull request by Septa2112**
- [x] I have read the [contributing guidelines](https://github.com/ggerganov/llama.cpp/blob/master/CONTRIBUTING.md)
- Self-reported review complexity:
  - [x] Low
  - [ ] Medium
  - [ ] High

---
### Changes
From https://github.com/ggerganov/llama.cpp/discussions/9037#discussioncomment-10381324 and some related comments in `perplexity.cpp`, I noticed that `logits_all` seems to be deprecated?

So I made the following changes:

1. By setting `llama_batch.logits`, avoid relying on `logits_all == true` when running in function `perplexity_v2` .
2. Completed a todo along the way:  rename `llama_batch.logits` to `llama_batch.output`.

---
### Test Platform

Linux 6.5.0-41-generic #41~22.04.2-Ubuntu SMP PREEMPT_DYNAMIC Mon Jun  3 11:32:55 UTC 2 x86_64 x86_64 x86_64 GNU/Linux

- CPU: i7-11700k

### Test Command
```bash
./build/bin/llama-perplexity -m ../gguf_models/llama-2-7b.Q2_K.gguf -f ../prompt/article.txt -t 16 --ppl-stride 220
```

#### Before
```
perplexity_v2: tokenizing the input ..
perplexity_v2: have 4394 tokens. Calculation chunk = 2176
perplexity_v2: calculating perplexity over 11 chunks, batch_size=2048
perplexity_v2: 68.55 seconds per pass - ETA 12.57 minutes
[1]1.0052,[2]1.0334,[3]1.0239,[4]1.0191,[5]1.0160,[6]1.3106,[7]1.5908,[8]1.5034,[9]1.4402,[10]1.3894,[11]1.3493,
 
llama_print_timings:        load time =     469.82 ms
llama_print_timings:      sample time =       0.00 ms /     1 runs   (    0.00 ms per token,      inf tokens per second)
llama_print_timings: prompt eval time =  807898.38 ms / 23936 tokens (   33.75 ms per token,    29.63 tokens per second)
llama_print_timings:        eval time =       0.00 ms /     1 runs   (    0.00 ms per token,      inf tokens per second)
llama_print_timings:       total time =  810532.89 ms / 23937 tokens
```

#### After
```
perplexity_v2: tokenizing the input ..
perplexity_v2: have 4394 tokens. Calculation chunk = 2176
perplexity_v2: calculating perplexity over 11 chunks, batch_size=2048
perplexity_v2: 68.83 seconds per pass - ETA 12.62 minutes
[1]1.0052,[2]1.0334,[3]1.0239,[4]1.0191,[5]1.0160,[6]1.3106,[7]1.5908,[8]1.5034,[9]1.4402,[10]1.3894,[11]1.3493,
 
llama_print_timings:        load time =     465.04 ms
llama_print_timings:      sample time =       0.00 ms /     1 runs   (    0.00 ms per token,      inf tokens per second)
llama_print_timings: prompt eval time =  795179.21 ms / 23936 tokens (   33.22 ms per token,    30.10 tokens per second)
llama_print_timings:        eval time =       0.00 ms /     1 runs   (    0.00 ms per token,      inf tokens per second)
llama_print_timings:       total time =  797794.73 ms / 23937 tokens
```

---
### Others
If these changes are acceptable to the community, I'd like to modify other places that depend on `logits_all == true`, such as
https://github.com/ggerganov/llama.cpp/blob/90db8146d56d83a605f2c475eca39bcda29cf16d/examples/perplexity/perplexity.cpp#L1797-L1801

https://github.com/ggerganov/llama.cpp/blob/90db8146d56d83a605f2c475eca39bcda29cf16d/examples/imatrix/imatrix.cpp#L516-L520

### CODE REVIEW

**Comment by Septa2112** [comment]
## Test Platform

Linux 6.5.0-41-generic #41~22.04.2-Ubuntu SMP PREEMPT_DYNAMIC Mon Jun  3 11:32:55 UTC 2 x86_64 x86_64 x86_64 GNU/Linux


## Test Command
```bash
./build/bin/llama-perplexity -m ../gguf_models/llama-2-7b.Q2_K.gguf -f ../prompt/article.txt -t 16 --ppl-stride 220
```

## Test Results
### Before
```
perplexity_v2: tokenizing the input ..
perplexity_v2: have 4394 tokens. Calculation chunk = 2176
perplexity_v2: calculating perplexity over 11 chunks, batch_size=2048
perplexity_v2: 68.55 seconds per pass - ETA 12.57 minutes
[1]1.0052,[2]1.0334,[3]1.0239,[4]1.0191,[5]1.0160,[6]1.3106,[7]1.5908,[8]1.5034,[9]1.4402,[10]1.3894,[11]1.3493,
 
llama_print_timings:        load time =     469.82 ms
llama_print_timings:      sample time =       0.00 ms /     1 runs   (    0.00 ms per token,      inf tokens per second)
llama_print_timings: prompt eval time =  807898.38 ms / 23936 tokens (   33.75 ms per token,    29.63 tokens per second)
llama_print_timings:        eval time =       0.00 ms /     1 runs   (    0.00 ms per token,      inf tokens per second)
llama_print_timings:       total time =  810532.89 ms / 23937 tokens
```

### After
```
perplexity_v2: tokenizing the input ..
perplexity_v2: have 4394 tokens. Calculation chunk = 2176
perplexity_v2: calculating perplexity over 11 chunks, batch_size=2048
perplexity_v2: 68.83 seconds per pass - ETA 12.62 minutes
[1]1.0052,[2]1.0334,[3]1.0239,[4]1.0191,[5]1.0160,[6]1.3106,[7]1.5908,[8]1.5034,[9]1.4402,[10]1.3894,[11]1.3493,
 
llama_print_timings:        load time =     465.04 ms
llama_print_timings:      sample time =       0.00 ms /     1 runs   (    0.00 ms per token,      inf tokens per second)
llama_print_timings: prompt eval time =  795179.21 ms / 23936 tokens (   33.22 ms per token,    30.10 tokens per second)
llama_print_timings:        eval time =       0.00 ms /     1 runs   (    0.00 ms per token,      inf tokens per second)
llama_print_timings:       total time =  797794.73 ms / 23937 tokens
```

File: examples/perplexity/perplexity.cpp [hunk 1]
```diff
@@ -406,13 +404,27 @@ static results_perplexity perplexity_v2(llama_context * ctx, const gpt_params &
             const int batch_start = start + j * n_batch;
             const int batch_size  = std::min(end - batch_start, n_batch);
 
+            llama_batch batch = llama_batch_init(batch_size, 0, 1);
```
Comment (compilade) [inline examples/perplexity/perplexity.cpp:407]: This can be allocated and freed outside of the loop. The same batch can be re-used with `llama_batch_clear(batch)`.

File: examples/perplexity/perplexity.cpp [hunk 2]
```diff
@@ -406,13 +404,27 @@ static results_perplexity perplexity_v2(llama_context * ctx, const gpt_params &
...
+                batch.token  [k] = tokens[idx];
+                batch.output [k] = 1;
+            }
+            batch.n_tokens   = batch_size;
+            batch.pos        = nullptr;
+            batch.n_seq_id   = nullptr;
+            batch.seq_id     = nullptr;
+            batch.all_pos_0  = j*n_batch;
+            batch.all_pos_1  = 1;
+            batch.all_seq_id = 0;
```
Comment (compilade) [inline examples/perplexity/perplexity.cpp:419]: This could also be done simply with

```c
llama_batch_add(batch, tokens[idx], j*n_batch + k, { 0 }, true);
```

within the loop over `k`.

**Comment by compilade** [comment]
@Septa2112

> If these changes are acceptable to the community, I'd like to modify other places that depend on logits_all == true, such as
>
> https://github.com/ggerganov/llama.cpp/blob/90db8146d56d83a605f2c475eca39bcda29cf16d/examples/imatrix/imatrix.cpp#L516-L520


Hi, I wrote that comment in #6122, but after studying `imatrix.cpp` more I think for `imatrix` specifically that always using all logits is fine, because otherwise the hidden state going through the FFN of the last layer would get trimmed, which would not be good for the importance matrix calculation.

Though it's true that it could avoid using `logits_all`, while still specifying that all logits should be output, this is already what is internally done when using `logits_all` (still, removing it *could* be worthwhile).

For the strided perplexity calculation (aka `perplexity_v2`) which is modified here, the main reason this was not updated to use an allocated `llama_batch` is because I was not sure *which* output should be kept. For the `perplexity()` function, it's simply the second half of the logits, while for this one, I'm not sure.

I see in <https://github.com/ggerganov/llama.cpp/pull/9102/commits/b0c6ad778d62fcf7dfda6ad88b1c24c1bb6b519f> that you've simply made it ask for all outputs, same as using `logits_all`. This is a good start because using an allocated `llama_batch` is necessary for more control over which outputs to keep.

If you do figure out which outputs to keep, know that the logits in the buffer returned by `llama_get_logits` are all contiguous (no holes). This means if outputs are excluded, then the offsets in that buffer will need to be corrected. `llama_get_logits_ith` abstracts this away and uses the original indices used in the `llama_batch`, but can only be expected to return a pointer to the logits of a single token at a time (although technically, this is all the same buffer, so it's also possible to use it simply to get the offset of a contiguous range).

Hopefully this helps!

File: examples/perplexity/perplexity.cpp (hunk 2 above)
Comment (Septa2112) [reply examples/perplexity/perplexity.cpp]: Changes have been made.

**Comment by Septa2112** [comment]
@compilade Thanks for your suggestion! I will carefully consider your advice and propose changes to `imatrix.cpp` in another separate PR. 
In this current PR, I only plan to modify the `perplexity_v2` section.
